install-gateway.ps1

```diff
@@ -1,79 +1,144 @@
+  }
+  
+  # Setting output filename 
+  if (!$outFile)
+  {
+    $outFile = [System.Net.WebUtility]::UrlDecode($uri.Split('/')[-1])
+  }
 
-# working directory path
-  $workd = "c:\temp"
+  $filename = "$workd\$outFile"
+  
+  write-Verbose "Downloading $uri into $filename"
+  (New-Object System.Net.WebClient).downloadFile($uri,$filename)
 
-# Check if work directory exists if not create it
-    If (!(Test-Path -Path $workd -PathType Container))
-    {
-      New-Item -Path $workd -ItemType directory
-    }
+  Print-FileInfo $filename
+
+  return $filename
+}
 
-#create config file for silent install
+function Install-JRE([string]$jresource = $script:jresource, [string]$workd = $script:workd)
+{
+# Download and silent install Java Runtime Environement
+# create config file for silent install
   $text = '
     INSTALL_SILENT=Enable
     AUTO_UPDATE=Enable
     SPONSORS=Disable
     REMOVEOUTOFDATEJRES=1
     '
-    $text | Set-Content "$workd\jreinstall.cfg"
+  $text | Set-Content "$workd\jreinstall.cfg"
 
-#download executable, this is the small online installer
-    $source = $jresource
-    $destination = "$workd\jreInstall.exe"
-    $client = New-Object System.Net.WebClient
-    $client.DownloadFile($source, $destination)
+# download executable, this is the small online installer
+  $filename = Download-File -uri $jresource -outFile 'jreInstall.exe'
 
-#install silently
-    Start-Process -FilePath "$workd\jreInstall.exe" -ArgumentList INSTALLCFG="$workd\jreinstall.cfg"
+# install silently
+  write-Verbose "Install JRE from $filename"
+  $process = Start-Process -FilePath "$filename" -ArgumentList "INSTALLCFG=`"$workd\jreinstall.cfg`"" -Wait -PassThru
 
-# Wait 120 Seconds for the installation to finish
-    Start-Sleep -s 180
+# Wait for the installation to finish
+  Write-Host 'Waiting for jreinstall.exe to finish...'
+  while ((Get-Process jreinstall -ErrorAction SilentlyContinue))
+  {
+    Start-Sleep -Seconds 10
+  }
 
 # Remove the installer
-    rm -Force $workd\jre*
+  rm -Force $workd\jre*
+
+  Print-ExecInfo $process
 }
 
-function Install-VcRedist()
+function Install-VcRedist([string]$vcrsource = $script:vcrsource, [string]$vcrsource_filter = $script:vcrsource_filter)
 {
-  $workd = "c:\temp"
-    If (!(Test-Path -Path $workd -PathType Container))
-    {
-      New-Item -Path $workd -ItemType directory
-    }
-  $destination = "$workd\vcredist_x64.exe"
-    $client = New-Object System.Net.WebClient
-    $client.DownloadFile("https://download.microsoft.com/download/3/2/2/3224B87F-CFA0-4E70-BDA3-3DE650EFEBA5/vcredist_x64.exe", $destination)
-    Start-Process -FilePath "$workd\vcredist_x64.exe" -ArgumentList "/Q" -Wait
+  $filename = Download-File -uri $vcrsource -msdl_filter $vcrsource_filter
+
+  Write-Verbose "Installing VcRedist from $filename"
+  $process = Start-Process -FilePath "$filename" -ArgumentList "/Q /log $workd\vcredist.log" -Wait -PassThru
+  
+  Start-Sleep 2
```

Comment:
I see that you made a neater process waiting in JRE install, can we do the same here please? Also some max timeout might be useful, so that we don't wait forever and fail after a certain time.